File: java/twoSumSolution.java
import java.util.Arrays;

/** Two Sum
 * Problem link: https://leetcode.com/problems/two-sum/
 * 
 * Idea:
 *  I believe I can iterate through the list items from each element to the last until I find the expected result
 * 
 * Complexity: O(n^2)
 */
public class twoSumSolution {
    public static int[] twoSum(int[] nums, int target) {
        for (int i = 0 ; i < nums.length ; i++) {
            if (i + 1 == nums.length) {
                int[] result = {i, i + 1};
                return result;
            }

            for (int j = i + 1 ; j < nums.length ; j++) {
                if (nums[i] + nums[j] == target) {
                    int[] result = {i, j};
                    return result;
                }
            }
        }

        return null;
    }

    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 15};

        int[] solution = twoSum(nums, 9);

        System.out.println("Solution: " + Arrays.toString(solution));
    }
}

File: python/twoSumSolution.py
''' Two Sum

Problem link: https://leetcode.com/problems/two-sum/

Idea: This time, I tried a different approach. When iterating through the list, I start from index 1. I had seen this solution before but didn't fully understand it until I applied it myself. Starting from index 1 makes the iteration more efficient by eliminating at least one unnecessary comparison. Additionally, this approach avoids use duplicated indices.

Complexity: O(n^2)
'''

from typing import List

def twoSum(nums: List[int], target: int) -> List[int]:
    m = {}
    for i,num in enumerate(nums):
        if target-num in m:
            return [i,m[target-num]]
        m[num] = i

print(f"Solution {twoSum([2, 7, 11, 15], 9)}")
print(f"Solution {twoSum([3, 2, 4], 6)}")
print(f"Solution {twoSum([3, 3], 6)}")